llvm/test/Transforms/InstCombine/commutative-operation-over-selects.ll

```diff
@@ -0,0 +1,1117 @@
+  %res = call {i16, i1} @llvm.umul.with.overflow.i16(i16 %s1, i16 %s0)
+  %obit = extractvalue {i16, i1} %res, 1
+  br i1 %obit, label %overflow, label %normal
+overflow:
+  ret i16 0
+normal:
+  %mul = extractvalue {i16, i1} %res, 0
+  ret i16 %mul
+}
+
+define i16 @fold_select_smul_fix_neg(i1 %c, i1 %c1, i16 %a, i16 %b, i32 %y) {
+; CHECK-LABEL: define i16 @fold_select_smul_fix_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], i16 [[A:%.*]], i16 [[B:%.*]], i32 [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i16 [[A]], i16 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], i16 [[B]], i16 [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call i16 @llvm.smul.fix.i16(i16 [[S1]], i16 [[S0]], i32 1)
+; CHECK-NEXT:    ret i16 [[RET]]
+;
+  %s0 = select i1 %c, i16 %a, i16 %b
+  %s1 = select i1 %c1, i16 %b, i16 %a
+  %ret = call i16 @llvm.smul.fix.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define i16 @fold_select_umul_fix_neg(i1 %c, i1 %c1, i16 %a, i16 %b, i32 %y) {
+; CHECK-LABEL: define i16 @fold_select_umul_fix_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], i16 [[A:%.*]], i16 [[B:%.*]], i32 [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i16 [[A]], i16 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], i16 [[B]], i16 [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call i16 @llvm.umul.fix.i16(i16 [[S1]], i16 [[S0]], i32 1)
+; CHECK-NEXT:    ret i16 [[RET]]
+;
+  %s0 = select i1 %c, i16 %a, i16 %b
+  %s1 = select i1 %c1, i16 %b, i16 %a
+  %ret = call i16 @llvm.umul.fix.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define i16 @fold_select_smul_fix_sat_neg(i1 %c, i1 %c1, i16 %a, i16 %b, i32 %y) {
+; CHECK-LABEL: define i16 @fold_select_smul_fix_sat_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], i16 [[A:%.*]], i16 [[B:%.*]], i32 [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i16 [[A]], i16 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], i16 [[B]], i16 [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call i16 @llvm.smul.fix.sat.i16(i16 [[S1]], i16 [[S0]], i32 1)
+; CHECK-NEXT:    ret i16 [[RET]]
+;
+  %s0 = select i1 %c, i16 %a, i16 %b
+  %s1 = select i1 %c1, i16 %b, i16 %a
+  %ret = call i16 @llvm.smul.fix.sat.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define i16 @fold_select_umul_fix_sat_neg(i1 %c, i1 %c1, i16 %a, i16 %b, i32 %y) {
+; CHECK-LABEL: define i16 @fold_select_umul_fix_sat_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], i16 [[A:%.*]], i16 [[B:%.*]], i32 [[Y:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i16 [[A]], i16 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], i16 [[B]], i16 [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = call i16 @llvm.umul.fix.sat.i16(i16 [[S1]], i16 [[S0]], i32 1)
+; CHECK-NEXT:    ret i16 [[RET]]
+;
+  %s0 = select i1 %c, i16 %a, i16 %b
+  %s1 = select i1 %c1, i16 %b, i16 %a
+  %ret = call i16 @llvm.umul.fix.sat.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define float @fold_select_fma_neg(i1 %c, i1 %c1, float %a, float %b, float %y) {
```

Comment:
Change this test to instead check the case where the selects are matching, but on arguments 2 and 3 instead of 1 and 2.